File: leetcode/2001-3000/2601-2700/2661-2670/2661-first-completely-painted-row-or-column/FirstCompletelyPaintedRowOrColumn.java
public class FirstCompletelyPaintedRowOrColumn {
  public static void main(String[] args) {
    int[] arr = {
        2, 8, 7, 4, 1, 3, 5, 6, 9
    };

    int[][] mat = {
        { 3, 2, 5 },
        { 1, 4, 6 },
        { 8, 7, 9 }
    };

    System.out.println(new Solution().firstCompleteIndex(arr, mat));
  }
}

class Solution {
  public int firstCompleteIndex(int[] arr, int[][] mat) {
    int[] rows = new int[mat.length];
    int[] columns = new int[mat[0].length];

    int[] values = new int[mat.length * mat[0].length];

    for (int i = 0; i < mat.length; i++) {
      for (int j = 0; j < mat[0].length; j++) {
        int index = i * mat[0].length + j;
        values[index] = mat[i][j];
      }
    }

    int[] indices = new int[values.length];

    for (int i = 0; i < values.length; i++) {
      indices[values[i] - 1] = i;
    }

    for (int i = 0; i < arr.length; i++) {
      int a = arr[i];
      int v = indices[a - 1];

      int r = v / mat[0].length;
      int c = v % mat[0].length;

      rows[r]++;
      columns[c]++;

      if (rows[r] == columns.length || columns[c] == rows.length) {
        return i;
      }
    }

    return 0;
  }
}

File: leetcode/2001-3000/2601-2700/2661-2670/2661-first-completely-painted-row-or-column/FirstCompletelyPaintedRowOrColumn.ts
function firstCompleteIndex(arr: number[], mat: number[][]): number {
  const m = mat.length;
  const n = mat[0].length;

  const rows = new Array(m).fill(n);
  const columns = new Array(n).fill(m);

  const values = new Array(m * n).fill(0);

  for (let i = 0; i < mat.length; i++) {
    for (let j = 0; j < mat[0].length; j++) {
      const index = i * n + j;

      values[index] = mat[i][j];
    }
  }

  const indices = new Array(values.length).fill(0);

  for (let i = 0; i < values.length; i++) {
    indices[values[i] - 1] = i;
  }

  for (let i = 0; i < arr.length; i++) {
    const a = arr[i];
    const v = indices[a - 1];

    // The division returns a float, so we need a floor here.
    const r = Math.floor(v / n);
    const c = v % n;

    rows[r]--;
    columns[c]--;

    if (rows[r] === 0 || columns[c] === 0) {
      return i;
    }
  }

  return 0;
};

// console.log(firstCompleteIndex([2, 8, 7, 4, 1, 3, 5, 6, 9], [[3, 2, 5], [1, 4, 6], [8, 7, 9]]));
console.log(firstCompleteIndex([6, 2, 3, 1, 4, 5], [[5, 1], [2, 4], [6, 3]]));
